### PR TITLE
[SYCL] Add size_type, to accessor and local_accessor classes.

### PR DESCRIPTION
**Pull request by mmoadeli**
- Add size_type to accessor and local_accessor classes.

### CODE REVIEW

File: sycl/include/sycl/accessor.hpp [hunk 1]
```diff
@@ -2972,6 +2974,12 @@ class __SYCL_EBO host_accessor
   }
 
 public:
+  using value_type = typename std::conditional<AccessMode == access_mode::read,
+                                               const DataT, DataT>::type;
+  using reference = value_type &;
+  using const_reference = const DataT &;
+  using size_type = size_t;
```
Comment (steffenlarsen) [inline sycl/include/sycl/accessor.hpp:2981]: `host_accessor` currently publicly inherit from `accessor<..., access::target::host_buffer>` which should define these. Do we really need these redefined?

**Review by steffenlarsen — COMMENTED** [review]
Could you please add a small in-tree LIT test making sure `size_type` is available for the various accessors and that it is the right type?

File: sycl/include/sycl/accessor.hpp (hunk 1 above)
Comment (mmoadeli) [reply sycl/include/sycl/accessor.hpp:2981]: Thanks @steffenlarsen 
It's added following the 
[Interface for buffer host accessors](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_interface_for_buffer_host_accessors) in specifications.

**Comment by mmoadeli** [comment]
> Could you please add a small in-tree LIT test making sure `size_type` is available for the various accessors and that it is the right type?

Thanks @steffenlarsen , done.

**Comment by bader** [comment]
@mmoadeli, OpenCL CPU and ESIMD emulator results were invalid due to infrastructure issue. I'll restart the pre-commit to get relevant results as soon as possible. No actions from your side are required. Sorry for inconvenience.

File: sycl/include/sycl/accessor.hpp (hunk 1 above)
Comment (steffenlarsen) [reply sycl/include/sycl/accessor.hpp:2981]: What I mean is, these aliases are defined in the base class (`accessor<..., access::target::host_buffer>`) of `host_accessor` and should already be defined through it. That is, they should be available on `host_accessor` even if they are not added here.

Comment (mmoadeli) [reply sycl/include/sycl/accessor.hpp:2981]: @steffenlarsen I understand your point. My impression was that there might have been a reasoning for having them there. I remove it and any other ones declared in base class, from the host_accessor.

Comment (steffenlarsen) [reply sycl/include/sycl/accessor.hpp]: At some point we'll have to remove the base class, given it is deprecated, but until then it is fine to rely on it. 😄

**Review by steffenlarsen — APPROVED** [review]
LGTM!

File: sycl/test/basic_tests/accessor/accessor_size_type.cpp [hunk 2]
```diff
@@ -0,0 +1,15 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN: %t.out
```
Comment (AlexeySachkov) [inline sycl/test/basic_tests/accessor/accessor_size_type.cpp:2]:
```suggestion
// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -fsyntax-only
```

There is nothing to run in this test

Comment (mmoadeli) [reply sycl/test/basic_tests/accessor/accessor_size_type.cpp]: @AlexeySachkov, right, thanks, done.

**Comment by bader** [comment]
Shouldn't we use C++ type (i.e. `std::size_t`) instead of C type (i.e. `size_t`)?

**Comment by mmoadeli** [comment]
> Shouldn't we use C++ type (i.e. `std::size_t`) instead of C type (i.e. `size_t`)?

Specification [Interface for buffer command accessors](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_interface_for_buffer_command_accessors) defines it as `size_t`.

**Comment by bader** [comment]
> > Shouldn't we use C++ type (i.e. `std::size_t`) instead of C type (i.e. `size_t`)?
> 
> Specification [Interface for buffer command accessors](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_interface_for_buffer_command_accessors) defines it as `size_t`.

Okay. Let me forward this question to @intel/dpcpp-specification-reviewers.

**Comment by gmlueck** [comment]
> > Shouldn't we use C++ type (i.e. `std::size_t`) instead of C type (i.e. `size_t`)?
> 
> Specification [Interface for buffer command accessors](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_interface_for_buffer_command_accessors) defines it as `size_t`.

I agree that the SYCL spec should use the `std` namespace when referring to `size_t`, and I opened an issue for this:

https://gitlab.khronos.org/sycl/Specification/-/issues/648

It would be better to use `std::size_t` in the CTS also.

FWIW, all C++ compilers that I know of recognize `size_t` (without namespace) as an alias to `std::size_t` even though the C++ spec does not guarantee this.

**Comment by AlexeySachkov** [comment]
```
********************
Unresolved Tests (1):
  SYCL :: AsyncHandler/default_async_handler.cpp

********************
Timed Out Tests (4):

3 warning(s) in tests
  SYCL :: KernelAndProgram/spec_constants_after_link.cpp
  SYCL :: Regression/kernel_bundle_ignore_sycl_external.cpp
  SYCL :: Regression/same_unnamed_kernels.cpp
  SYCL :: Regression/static-buffer-dtor.cpp
```

Last 4 tests were disabled on HIP in https://github.com/intel/llvm-test-suite/pull/1639, `async_handler` is reported in #8555

**Pull request merged**